File: .leetcode/376.摆动序列.cpp
/*
 * @lc app=leetcode.cn id=376 lang=cpp
 *
 * [376] 摆动序列
 */
#include<bits/stdc++.h>
using namespace std;

// @lc code=start
class Solution {
public:
    int wiggleMaxLength(vector<int>& nums) {
        if(nums.size()<2)return nums.size();
        int pre=0,cur=0,res=2;
        for(int i = 1;i < nums.size();i++){
            cur = nums[i] - nums[i-1];
            if(cur*pre<0)res++;
            if(!cur)cur = pre;
            pre = cur;
        }
        if(cur==pre&&pre==0&&res==2&&nums[0]==nums[nums.size()-1])return 1;
        return res;
    }
};
// @lc code=end

File: .leetcode/376.摆动序列.java
/*
 * @lc app=leetcode.cn id=376 lang=java
 *
 * [376] 摆动序列
 */

// @lc code=start
class Solution {
    public int wiggleMaxLength(int[] nums) {
        int n = nums.length;
        if(n == 0 || n == 1)return n;
        int pre = 0;
        int ans = 1;
        for(int i = 0;i < n - 1;i++){
            int cur = nums[i + 1] - nums[i];
            if((cur > 0 && pre <= 0) || (cur < 0 && pre >= 0)){
                ans++;
                pre = cur;
            }
        }
        return ans;
    }
}
// @lc code=end

